RPP: Added error states

zmaglica opened the pull request:
Fixes #7425 

#### Changes proposed in this Pull Request

This PR introduces error states and their management. It ensures that all exceptions, including minor ones, are effectively handled. The core principle is to avoid returning exceptions from states. Instead, states should consistently return their designated state. This approach eliminates the need for exception handling in services utilizing these states or in the gateway, where they were previously handled.

Error states:

- `AbstractPaymentErrorState` : Abstract error class that holds common or reusable logic for all error states
- `PaymentErrorState`: Error state that is used when an error occurs on the buyer's side, which cannot be fixed through code.
- `PaymentRequestErrorState`: Handles `PaymentRequestException` that can occur when context is populated from the request.
- `SystemErrorState`: Equivalent of \Exception we have in PHP but used for our state machine. Look at this state as a default error state when an exception or any unwanted flow occurs, where we don't have a specific error state to transition.
- `WooPaymentsApiServerErrorState`: Error state that is used when server API throws an exception. I used WooPayments prefix to make it more clearer that this is the error state for API server errors on WooCommerce Payments server.



<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Make sure that all exceptions are handled by the error states. In other words, if an exception is triggered in state, the state needs to handle it and return error state. The only exception that can be handled without the error state is `StateTransitionException` which can occur outside state processing or within the catch block, where we transition from state to new error state.
* All error states should extend the base or abstract error class.
* Make sure that all error states have their logging and mark order as failed function configured property, and according to the business logic. What that means is that each error state can have disabled logging or disabled marking order as failed.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

zmaglica committed:
Handled Order not found exception

zmaglica committed:
Handled Amount too small exception

zmaglica committed:
handled StateTransitionException

zmaglica committed:
Handled Amount_Too_Small_Exception

zmaglica committed:
Handled PaymentRequestException

zmaglica committed:
Added api server error state

zmaglica committed:
Added abstract payment error state

zmaglica committed:
Added abstract error class to error state classes

zmaglica committed:
Fixed broken test

zmaglica committed:
Added missing test

zmaglica committed:
Added tests for should_* methods

zmaglica committed:
Added missing tests for error states

zmaglica committed:
Added tests for payment context

zmaglica committed:
Handled all exceptions in initial state

zmaglica committed:
Added tests for order service

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7739` or branch name `rpp-7425-handle-error-states` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=rpp-7425-handle-error-states) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 64683232977c02bc6012f133a10c63faa4e5c880
- Build time: 2023-11-21 13:23:57 UTC

_Note: the build is updated when a new commit is pushed to this PR._

jessy-p commented on src/Internal/Payment/State/AbstractPaymentErrorState.php:
```diff
@@ -0,0 +1,91 @@
+	 *
+	 * @return bool True if the error should be logged, otherwise false.
+	 */
+	protected function should_log_error(): bool {
```
Shouldn't all errors be logged by default?

zmaglica replied:
I think they need to, but I wasn't 100% certain, so I made it to false. We can easily set that all error states needs to be logged, by the default.

jessy-p reviewed (commented):
Under testing instructions, it would be nice to have a listing of the errors and maybe a one liner of the scenarios to test.

zmaglica commented:
> Under testing instructions, it would be nice to have a listing of the errors and maybe a one liner of the scenarios to test.

I have updated the test instructions and added a small description to each error state.

zmaglica committed:
Merge branch 'develop' into rpp-7425-handle-error-states

zmaglica committed:
Fixed psalm errors

zmaglica committed:
Added changelog

RadoslavGeorgiev commented on src/Internal/Payment/PaymentContext.php:
```diff
@@ -37,6 +37,13 @@ class PaymentContext {
 	 */
 	private $transitions = [];
 
+	/**
+	 * Stores exception that is triggered before error state is created.
+	 *
+	 * @var \Exception
+	 */
+	private $exception;
```
Like request classes, there is no need to define a property of the object: It's all stored within `data`. With that in mind, this one is not used, and can be removed.

RadoslavGeorgiev commented on src/Internal/Payment/PaymentContext.php:
```diff
@@ -300,6 +307,24 @@ public function get_transitions(): array {
+	 *
+	 * @param \Exception $exception Exception to store.
+	 */
+	public function set_exception( $exception ) {
```
I'm not insisting on this, but I'd prefer to call those `set_last_exception` and `get_last_exception`, similarly to how there's a `$wpdb->last_error`.

Also, please add a proper type hint to the method:
```suggestion
	public function set_exception( Exception $exception ) {
```

Similar,y `get_(last_)exception()` should return `: ?Exception`

RadoslavGeorgiev commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -106,49 +106,44 @@ public function __construct(
+				return $duplicate_payment_result;
+			}
+
+			/**
+			 * Payments are based on intents, and intents use customer objects for billing details.
+			 *
+			 * The customer is created/updated right before requesting the creation of
+			 * a payment intent, and the two actions must be adjacent to each-other.
+			 */
```
This docblock is completely out of place now.

RadoslavGeorgiev commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -161,23 +156,34 @@ public function start_processing( PaymentRequest $request ) {
+			return $next_state->complete_processing();
+		} catch ( Amount_Too_Small_Exception $e ) {
+			$this->minimum_amount_service->store_amount_from_exception( $e );
+			return $this->create_error_state( SystemErrorState::class, $e );
```
The amount being too small is not a system exception, it happens whenever the amount to pay is less than the required minimum. This is not an indication that there's something wrong with the system, just a payment error.

RadoslavGeorgiev commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -161,23 +156,34 @@ public function start_processing( PaymentRequest $request ) {
+			$next_state = $this->create_state( ProcessedState::class );
 
-		return $next_state->complete_processing();
+			return $next_state->complete_processing();
```
By positioning this line within the `catch` block of this state, we're essentially trying to handle all exceptions from the `ProcessedState` here as well. This goes against the idea of granularly handling exceptions.

RadoslavGeorgiev commented on src/Internal/Payment/State/AbstractPaymentErrorState.php:
```diff
@@ -0,0 +1,91 @@
+		throw new Exception( __( 'The payment process could not be completed.', 'woocommerce-payments' ) );
+	}
+
+	/**
+	 * Determines whether an error should be logged.
+	 *
+	 * @return bool True if the error should be logged, otherwise false.
+	 */
+	protected function should_log_error(): bool {
+		return false;
+	}
```
Logging of the whole context is already built, and specific errors (ex. API errors) are logged already. I'd remove the logging functionality from erroneous states altogether.

RadoslavGeorgiev reviewed (changes requested):
I have not reviewed all the details regarding the PR, but I think we need a significant change in direction. Keep this line from the issue in mind:

> The new payment process should handle exceptions **at a more granular level**, and switch to a failure state immediately.

The current implementation in the PR goes against granularity, as we're back to the same state as the current payment process, where all exceptions are caught in the same place.

A few notes:
- `StateTransitionException`s were a placeholder for exceptions in specific scenarios. We didn't want to create too many separate exception classes, in case they needed to be replaced by this stream. However, all but the `create_state()`'s `StateTransitionException`s should be replaced by something else.
- Please merge `develop` into the PR, there are some more exceptions after Dat merged some PRs.

Presently we're not storing the context anywhere, but keep this question in mind: If we were to store payments, how would we approach erroneous states when we need to retry those payments?

dmvrtx commented:
There has been no progress on this pull request since November 2023. While it suggests an approach to improve the payment flow, it is now outdated with numerous conflicts. I'm closing this pull request. If the issue will be worked on, it should be done anew.
